src/librustc_resolve/error_reporting.rs

```diff
@@ -564,10 +565,291 @@ impl<'a, 'b:'a> ImportResolver<'a, 'b> {
+                    ) {
+                        debug!("check_for_module_export_macro: previous_span={:?}", previous_span);
+                        removal_span = removal_span.with_lo(previous_span.lo());
+                    }
+                }
+                debug!("check_for_module_export_macro: removal_span={:?}", removal_span);
+
+                // Find the span after the crate name and if it has nested imports immediatately
+                // after the crate name already.
+                //   ie. `use a::b::{c, d};`
+                //               ^^^^^^^^^
+                //   or  `use a::{b, c, d}};`
+                //               ^^^^^^^^^^^
+                let (has_nested, after_crate_name) = find_span_immediately_after_crate_name(
+                    self.resolver.session, module_name, directive.use_span,
+                );
+                debug!("check_for_module_export_macro: has_nested={:?} after_crate_name={:?}",
+                       has_nested, after_crate_name);
+
+                let source_map = self.resolver.session.source_map();
+
+                // Remove two bytes at the end to keep all but the `};` characters.
+                //   ie. `{b::{c, d}, e::{f, g}};`
+                //        ^^^^^^^^^^^^^^^^^^^^^
+                let end_bytes = BytePos(if has_nested { 2 } else { 1 });
+                let mut remaining_span = after_crate_name.with_hi(
+                    after_crate_name.hi() - end_bytes);
+                if has_nested {
+                    // Remove two bytes at the start to keep all but the initial `{` character.
+                    //   ie. `{b::{c, d}, e::{f, g}`
+                    //         ^^^^^^^^^^^^^^^^^^^^
+                    remaining_span = remaining_span.with_lo(after_crate_name.lo() + BytePos(1));
+                }
+
+                // Calculate the number of characters into a snippet to remove the removal
+                // span.
+                let lo = removal_span.lo() - remaining_span.lo();
+                let hi = lo + (removal_span.hi() - removal_span.lo());
+                if let Ok(mut remaining) = source_map.span_to_snippet(remaining_span) {
+                    // Remove the original location of the binding.
+                    remaining.replace_range((lo.0 as usize)..(hi.0 as usize), "");
+                    correction = format!("use {}::{{{}, {}}};", module_name, import, remaining);
+                }
+            }
+
+            let suggestion = Some((
+                span,
+                String::from("a macro with this name exists at the root of the crate"),
+                correction,
+                Applicability::MaybeIncorrect,
+            ));
+            let note = vec![
+                "this could be because a macro annotated with `#[macro_export]`...".to_string(),
+                "...will be exported at the root of the crate instead of the...".to_string(),
+                "...module where it is defined".to_string(),
```

Comment:
Does this really need to be three separate lines? We haven't shied away from long notes in other errors, only splitting them in this way when they can potentially have spans.

Reply:
Fixed.